src/backend/optimizer/path/allpaths.c

```diff
@@ -2135,6 +2135,12 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 		 */
 		subplan = share_prepared_plan(cteroot, cteplaninfo->shared_plan);
 		subroot = cteplaninfo->subroot;
+
+		if(subplan->flow->locustype == CdbLocusType_General)
+		{
+			subplan->flow->locustype = CdbLocusType_SingleQE;
+			subplan->flow->flotype = FLOW_SINGLETON;
+		}
```

Comment:
>Moving it completely to segments may lead to duplication of tuples.

How did you get that? Provide an example, please. And again, I have doubts that we have to always make General be SingleQE, because there are the cases, when it's safe to have General scan at all segments:
```
set gp_cte_sharing=1;
set optimizer=0;
create table t(i int, j int);
explain (costs off)
with cte as ( select i, j from (values (1,1),(2,2)) s(i,j)) select * from cte a, t, cte b where a.i=t.j and b.i=t.i;
                        QUERY PLAN                        
----------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Hash Join
         Hash Cond: (t.i = share0_ref1.i)
         ->  Hash Join
               Hash Cond: (t.j = share0_ref2.i)
               ->  Seq Scan on t
               ->  Hash
                     ->  Shared Scan (share slice:id 1:0)
         ->  Hash
               ->  Shared Scan (share slice:id 1:0)
                     ->  Materialize
                           ->  Values Scan on "*VALUES*"
 Optimizer: Postgres query optimizer
(13 rows)
```
OFC, the example above has Shared Scan is not cross-slice, but anyway.
Therefore, I guess, that the problem is mostly for Bottleneck join case.

Reply:
> Moving it completely to segments may lead to duplication of tuples.

if make the flow type to FLOW_PARTITIONED, thereby transferring sharedscan to segments. we will get that in the left subtree of the top join, each segment will produce a tuple and pass them to the coordinator, which will lead to duplication.

Reply:
> PR is trying to solve two problems, not one.

I still don't get it. Why Shared Scan can't have General locus?

Reply:
>Therefore, I guess, that the problem is mostly for Bottleneck join case.

technically, if we remove the two-phase limit, we will get a join on the segments and everything will be fine, since it is a place where singleQE join is created. but at the time of creating the two-phase limit, we don't know anything about shared scan.

Reply:
> if make the flow type to FLOW_PARTITIONED, thereby transferring sharedscan to segments. we will get that in the left subtree of the top join, each segment will produce a tuple and pass them to the coordinator, which will lead to duplication.

Provide example, please

Reply:
change from:
```
subplan->flow->locustype = CdbLocusType_SingleQE;
subplan->flow->flotype = FLOW_SINGLETON;
```
to
```
subplan->flow->locustype = CdbLocusType_Strewn;
subplan->flow->flotype = FLOW_PARTITIONED;
```
and you will get as many rows as segments as a result.
```
 Gather Motion 3:1  (slice5; segments: 3)
   ->  Hash Join
         Hash Cond: (b.c1 = share0_ref1.c1)
         ->  Redistribute Motion 1:3  (slice3; segments: 1)
               Hash Key: b.c1
               ->  Subquery Scan on b
                     ->  Limit
                           ->  Gather Motion 3:1  (slice2; segments: 3)
                                 ->  Limit
                                       ->  Hash Join
                                             Hash Cond: (d.c1 = share0_ref2.c1)
                                             ->  Seq Scan on d
                                             ->  Hash
                                                   ->  Redistribute Motion 3:3  (slice1; segments: 3)
                                                         Hash Key: share0_ref2.c1
                                                         ->  Shared Scan (share slice:id 1:0)
         ->  Hash
               ->  Redistribute Motion 3:3  (slice4; segments: 3)
                     Hash Key: share0_ref1.c1
                     ->  Shared Scan (share slice:id 4:0)
                           ->  Materialize
                                 ->  Aggregate
                                       ->  Values Scan on "*VALUES*"
```

Reply:
> I still don't get it. Why Shared Scan can't have General locus? Producer can generate dublicates?

I meant that transferring sharedscan to segments cannot be a solution. this cannot be in the current implementation.

Reply:
In what cases General locus will be redistirbuted to locus Hashed (without code changes)?

And repeate, please, what are the 2 problems that are solved by changing General to SingleQE?

Reply:
> by changing General to SingleQE

PR solves only one problem: the inability to build a valid plan.

Reply:
> In what cases General locus will be redistirbuted to locus Hashed (without code changes)?

could it be? because how can we distribute General if we don't know by which key (column) to do the distribution?

Reply:
> could it be? because how can we distribute General if we don't know by which key (column) to do the distribution?

Well, if not, when dublicates are possible (without code changes)?

Reply:
this can't happen.

Reply:
> technically, if we remove the two-phase limit, we will get a join on the segments and everything will be fine, since it is a place where singleQE join is created. but at the time of creating the two-phase limit, we don't know anything about shared scan.

I'll invesigte it in more details.